### PR TITLE
[WIP] Implement responsefiles support PR7050

### PR DESCRIPTION
**Pull request by bschommer**
Follow up to PR #748 against trunk.

### CODE REVIEW

**Comment by gasche** [comment]
This PR implements functions for parsing of argument files. Do you plan to eventually add commits that combine this with the new `Expand` support to actually provide `-args/args0` arguments to the compiler tools, or would that be out of the scope of the current PR?

**Comment by bschommer** [comment]
I added the functions to the compiler command line arguments.

File: driver/main_args.ml [hunk 1]
```diff
@@ -724,6 +724,19 @@ let mk_no_strict_formats f =
   \      and instead fix invalid formats.)"
 ;;
 
+let mk_expand_responsefile f =
```
Comment (damiendoligez) [inline driver/main_args.ml]: Please respect the format: these functions have the same name as the options they describe, with dashes replaced by underscores, so this should be `mk_args` and the next one should be `mk_args0`.

File: driver/main_args.ml [hunk 2]
```diff
@@ -813,6 +826,9 @@ module type Compiler_options = sig
 
   val _nopervasives : unit -> unit
   val _dtimings : unit -> unit
+
+  val _expand_responsefile: string -> string array
```
Comment (damiendoligez) [inline driver/main_args.ml]: Same remark about the function names.

File: stdlib/arg.ml [hunk 3]
```diff
@@ -336,3 +336,46 @@ let align ?(limit=max_int) speclist =
...
+  let stash inw =
+    if inw then begin
+      words := Buffer.contents buf :: !words;
+      Buffer.clear buf;
+    end in
+  let rec read inw =
+    try
+      let c = input_char ic in
+      if c = sep then begin
+        stash inw; read true
```
Comment (damiendoligez) [inline stdlib/arg.ml]: This should be `read false`, right?
In fact you should drop `inw` altogether because we decided we want to allow empty arguments. Just add a test at EOF to stash only when the buffer is not empty.

File: stdlib/arg.ml [hunk 4]
```diff
@@ -336,3 +336,46 @@ let align ?(limit=max_int) speclist =
...
+
+let write_aux sep file args =
+  let oc = open_out_bin file in
+  let first = ref true in
+  let sep () =
+    if !first then
+      first:=false
+    else
+      output_char oc sep in
+  Array.iter (fun s -> sep (); output_string oc s) args;
```
Comment (damiendoligez) [inline stdlib/arg.ml]: Get rid of the ugly `first` reference and output a "separator" after each arg. Formally, they are terminators, with the last one optional.

File: stdlib/arg.mli [hunk 5]
```diff
@@ -173,3 +173,19 @@ val current : int ref
     {!Arg.parse} uses the initial value of {!Arg.current} as the index of
     argument 0 (the program name) and starts parsing arguments
     at the next element. *)
+
+val read_arg: string -> string array
+(** [Arg.read_arg file] reads new line separated command line arguments from
```
Comment (damiendoligez) [inline stdlib/arg.mli]: According to your implementation, they are not "new line separated" but "linefeed-separated". What happened to the suggestion of ignoring CR in this case?

Also, this should be `terminated` rather than `separated`.

File: stdlib/arg.mli [hunk 6]
```diff
@@ -173,3 +173,19 @@ val current : int ref
...
+    file [file]. *)
+
+val read_arg0: string -> string array
+(** Identical to {!Arg.read_arg} but assumes NUL separated command line
```
Comment (damiendoligez) [inline stdlib/arg.mli]: Same remark about `separated` vs `terminated`.

File: stdlib/arg.mli [hunk 7]
```diff
@@ -173,3 +173,19 @@ val current : int ref
...
+    arguments *)
+
+val write_arg: string -> string array -> unit
+(** [Arg.write_arg file args] writes the arguments [args] newline separated into
```
Comment (damiendoligez) [inline stdlib/arg.mli]: terminated

File: stdlib/arg.mli [hunk 8]
```diff
@@ -173,3 +173,19 @@ val current : int ref
...
+    the file [file]. If the any of the arguments in [args] contains a newline
+    use the function {!Args.write_arg0} instead *)
+
+val write_arg0: string -> string array -> unit
+(** Identical to {!Arg.write_arg} but uses NUL as separator instead *)
```
Comment (damiendoligez) [inline stdlib/arg.mli]: terminator

File: testsuite/tests/lib-arg/testarg.ml [hunk 9]
```diff
@@ -124,6 +124,32 @@ let test_arg args = test spec (ref args);;
...
+      Printf.printf "%20s %c %-20s\n%!" x (if x = y then '=' else '#') y
+  in
+  Array.iter2 f argv argv';
+  Array.iter2 f argv argv0;
+  safe_rm "test_rw";
+  safe_rm "test_rw0";
+;;
+
+test_rw args1;;
+test_rw args2;;
```
Comment (damiendoligez) [inline testsuite/tests/lib-arg/testarg.ml:151]: You should also test with an empty argument list and a list that contains empty arguments.

**Review by damiendoligez — CHANGES_REQUESTED** [review]
I've added some change suggestions.

About the scope: I think it makes sense to add the options to all tools.

File: testsuite/tests/lib-arg/testarg.ml (hunk 9 above)
Comment (bschommer) [reply testsuite/tests/lib-arg/testarg.ml:151]: Is added.

File: stdlib/arg.mli (hunk 8 above)
Comment (bschommer) [reply stdlib/arg.mli]: Is changed.

File: stdlib/arg.mli (hunk 7 above)
Comment (bschommer) [reply stdlib/arg.mli]: Is changed.

File: stdlib/arg.mli (hunk 6 above)
Comment (bschommer) [reply stdlib/arg.mli]: Is changed.

File: stdlib/arg.mli (hunk 5 above)
Comment (bschommer) [reply stdlib/arg.mli]: Is changed.

File: stdlib/arg.ml (hunk 4 above)
Comment (bschommer) [reply stdlib/arg.ml]: Is changed.

File: stdlib/arg.ml (hunk 3 above)
Comment (bschommer) [reply stdlib/arg.ml]: Is changed.

File: driver/main_args.ml (hunk 2 above)
Comment (bschommer) [reply driver/main_args.ml]: Is changed.

File: driver/main_args.ml (hunk 1 above)
Comment (bschommer) [reply driver/main_args.ml]: Is changed.

**Comment by bschommer** [comment]
For the other tools: I will have a look but as @diml noted ocamldebug for example needs special care.

**Comment by bschommer** [comment]
I added ocamldep to the list of supported tools. I also added a simple wrapper around `parse_and_expand_argv_dynamic` that also prints error/help messages to `Arg`. In order to avoid the problems with `current` and the changed array the function does not modify `Arg.current` is not modified.

**Comment by damiendoligez** [comment]
> For the other tools: I will have a look but as @diml noted ocamldebug for example needs special care.

I did not mean to push hard for supporting all tools, but since you asked about scope (in https://github.com/ocaml/ocaml/pull/748#issuecomment-250517415) I wanted to give a positive answer. I'd be happy with a partial implementation anyway.

**Review by damiendoligez — APPROVED** [review]
Looks OK to me now.

**Comment by bschommer** [comment]
List of missing tools/programs:
- ocaml
- ocamldebug
- ocamlyacc 
- ocamlex
- ocamldoc
- cvt_emit
- cmpbyt
- ocamlmklib
- ocamlmktop
- stripdebug

**Comment by bschommer** [comment]
For stripdebug, cvt_emit, cmpbyt and ocamlyacc, ocamllex it does not make sense to add it since they expect a constant number of options.

ocamlmktop only passes the commands on.

ocamldebug, ocaml, ocamldoc and ocamlmklib all have either hand written parse functions or require some preprocessing of the arguments and I'm unsure if it works for them without problem.

**Comment by bschommer** [comment]
> I did not mean to push hard for supporting all tools, but since you asked about scope (in #748 (comment)) I wanted to give a positive answer. I'd be happy with a partial implementation anyway.

No problem, most of them are pretty straightforward.

**Comment by gasche** [comment]
I think that it would be nice to have support in `ocaml` and `ocamlnat` as well (you may have noticed that I mistakenly advertised it to Coq developers; I thought that the functorized argument-parsing machinery would provide the feature for the toplevels as well). It could be useful for the Coq use-case, as it allows as a debugging technique to "Drop" to the OCaml toplevel.

(Command-line is in the files `toplevel/topmain.ml` (for ocaml) and `toplevel/opttopmain.ml` (for ocamlnat). The native toplevel is not officially supported but it's on the path to redemption these days, so it's nice to keep it in synch.)

**Comment by bschommer** [comment]
I will have a look at ocaml and ocamlnat.

**Comment by bschommer** [comment]
It seems like it is not possible without some unsafe operations:

```
let override_sys_argv args =
  let len = Array.length args in
  if Array.length Sys.argv < len then invalid_arg "Toploop.override_sys_argv";
  Array.blit args 0 Sys.argv 0 len;
  Obj.truncate (Obj.repr Sys.argv) len;
  Arg.current := 0
```

**Comment by gasche** [comment]
The toplevel uses a horrible hack so that the first argument is the name of the executed file instead of the toplevel binary (I suppose). But how does this prevent handling `Expand` nodes in the argument array after this hack has been applied? I don't understand why there is a conflict here.

**Comment by bschommer** [comment]
`override_sys_argv` actually copies the rest of the un-parsed arguments into `Sys.argv` and then truncates `Sys.argv`. However `Expand` may increase the array beyond the length of `Sys.argv` and the copy would only work if the size is increased instead of making it smaller.

**Comment by gasche** [comment]
So if I understand correctly, the toplevel parses the provided argument list, then it stops on the first anonymous argument that is a source file (`.ml` extension), and considers that the remaining arguments are to be processed by the script itself upon execution, so it rewrites `Sys.argv` in place to have just the remaining argument as the control is passed to the script.

As it rewrites the arguments, it checks that the new argument-list is smaller than the previous one (as it is supposed to be a suffix of it). This could fail in general if one of the Expand node would generate a source file argument. (Note that if the Expand nodes don't generate source files argument, this would succeed, as the suffix would remain the same).

I'm confident there is a reasonable solution (for example we could just disable the size check that is certainly there for debugging), but it also seems to require a bit more work than I hoped. Would you agree with merging the current PR, and eventually support the toplevel in a following PR?

(I'm not implying that you personally have to do the toplevel-related work. If I have time I could try to implement it myself. Let us know what you want to do.)

**Comment by bschommer** [comment]
That is fine with me.

**Comment by gasche** [comment]
Could you add a Changes entry? Suggestion:

```
### Compiler user-interface and warnings:

[...]

- MPR#7050, GPR#748 GPR#843: new `-arg/-arg0 <file>` parameters to provide
  extra command-line arguments in a file -- see documentation.
  User programs may implement similar options
  using the new `Arg.Expand` constructor.
  (Bernhard Schommer, review by Jérémie Dimino, Gabriel Scherer
   and Damien Doligez, discussion with Alain Frisch and Xavier Leroy,
   feature request from the Coq team)
```

**Comment by gasche** [comment]
Thanks a lot for your hard work! Do you plan to look at the toplevel?

**Comment by bschommer** [comment]
I will have a look at it. Basically it seems that instead of truncate one needs some kind of adjust primitive. When I find a solution I will open a new PR.

**Comment by gasche** [comment]
(I now understand that relaxing the check is not enough as `truncate` actually requires a smaller value. Maybe the toplevel could just forbid source script names from within the return of `Expand`?)

**Comment by gasche** [comment]
@alainfrisch rightly pointed out that this feature was not documented (it's not in the manual or in the compiler manpages). I am planning on opening a Mantis ticket to track that need (tomorrow). @bschommer, would you be willing to send a patch for that?

File: stdlib/arg.ml [hunk 10]
```diff
@@ -336,3 +346,48 @@ let align ?(limit=max_int) speclist =
...
+  in
+  let rec read () =
+    try
+      let c = input_char ic in
+      if c = sep then begin
+        stash (); read ()
+      end else begin
+        Buffer.add_char buf c; read ()
+      end
+    with End_of_file ->
```
Comment (alainfrisch) [inline stdlib/arg.ml:375]: FWIW, this try...with breaks tail-recursion and would lead to stack overflow if the file is large enough.  Will submit a PR to address that shortly.
